src/main/java/frc/robot/subsystems/vision/Vision.java

```diff
@@ -0,0 +1,95 @@
+package frc.robot.subsystems.vision;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.photonvision.EstimatedRobotPose;
+import org.photonvision.PhotonCamera;
+import org.photonvision.PhotonPoseEstimator;
+import org.photonvision.PhotonPoseEstimator.PoseStrategy;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout.OriginPosition;
+import edu.wpi.first.apriltag.AprilTagFields;
+import edu.wpi.first.math.geometry.Transform3d;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Constants.VisionConstants;
+import frc.robot.Constants.FieldConstants;;
+
+/**
+ * Runnable that gets AprilTag data from PhotonVision.
+ */
+public class Vision implements Runnable {
+
+  private final PhotonPoseEstimator photonPoseEstimator;
+  private final PhotonCamera photonCamera;
+  private final AtomicReference<EstimatedRobotPose> atomicEstimatedRobotPose = new AtomicReference<EstimatedRobotPose>();
+
+  public Vision(PhotonCamera cameraName, Transform3d robotToCamera) {
+    this.photonCamera = cameraName;
+    PhotonPoseEstimator photonPoseEstimator = null;
+
+    try {
+      var layout = AprilTagFields.k2024Crescendo.loadAprilTagLayoutField();
+      layout.setOrigin(OriginPosition.kBlueAllianceWallRightSide); // Sets the Origin to the Blue AllianceWall and will be flipped in the robot thread
+
+      // if (photonCamera != null) {
+      photonPoseEstimator = photonCamera != null
+          ? new PhotonPoseEstimator(layout, PoseStrategy.MULTI_TAG_PNP_ON_COPROCESSOR, photonCamera, robotToCamera)
+          : null;
+      photonPoseEstimator.setMultiTagFallbackStrategy(PoseStrategy.LOWEST_AMBIGUITY);
+      // }
+    } catch (Exception e) {
+      DriverStation.reportError("Failed to load AprilTagFieldLayout", e.getStackTrace());
+      // photonPoseEstimator = null;
+    }
+
+    this.photonPoseEstimator = photonPoseEstimator;
+  }
+
+  @Override
+  public void run() {
+    // Get AprilTag data and updating the pose estimator
+    try {
+      if (photonPoseEstimator != null && photonCamera != null) {
+        var photonResults = photonCamera.getLatestResult(); //Gets the latest camera results
+        SmartDashboard.putString(photonCamera.getName(),
+            photonCamera.getLatestResult().getMultiTagResult().estimatedPose.toString());
+        if (photonResults.hasTargets() && (photonResults.targets.size() > 1
+            || photonResults.targets.get(0).getPoseAmbiguity() < VisionConstants.APRILTAG_AMBIGUITY_THRESHOLD)) {
```

Comment:
This condition would OK (will be true) if we have two targets both of which have big ambiguity? Is this what we want?

Reply:
This is per target, multiple camera calculation happen in the other file.